sdks/java/core/src/main/java/org/apache/beam/sdk/io/DefaultFilenamePolicy.java

```diff
@@ -72,52 +74,106 @@
   private static final Pattern SHARD_FORMAT_RE = Pattern.compile("(S+|N+|W|P)");
 
   /**
-   * Constructs a new {@link DefaultFilenamePolicy}.
    *
-   * @see DefaultFilenamePolicy for more information on the arguments to this function.
    */
-  @VisibleForTesting
-  DefaultFilenamePolicy(ValueProvider<String> prefix, String shardTemplate, String suffix) {
-    this.prefix = prefix;
-    this.shardTemplate = shardTemplate;
-    this.suffix = suffix;
+  public static class Params {
+    private ValueProvider<ResourceId> baseFilename;
+    private String shardTemplate;
+    private String suffix;
+
+    public Params(ValueProvider<ResourceId> baseFilename, String shardTemplate, String suffix) {
+      this.baseFilename = baseFilename;
+      this.shardTemplate = shardTemplate;
+      this.suffix = suffix;
+    }
+
+    public Params(ResourceId baseFilename, String shardTemplate, String suffix) {
+      this(StaticValueProvider.of(baseFilename), shardTemplate, suffix);
+    }
+
+    /**
+     * A helper function to construct a {@link DefaultFilenamePolicy.Params} using the standard
+     * filename parameters, namely a provided {@link ResourceId} for the output prefix, and
+     * possibly-null shard name template and suffix.
+     *
+     * <p>Any filename component of the provided resource will be used as the filename prefix.
+     *
+     * <p>If provided, the shard name template will be used; otherwise {@link
+     * #DEFAULT_UNWINDOWED_SHARD_TEMPLATE} will be used for non-windowed file names and {@link
+     * #DEFAULT_WINDOWED_SHARD_TEMPLATE} will be used for windowed file names.
+     *
+     * <p>If provided, the suffix will be used; otherwise the files will have an empty suffix.
+     */
+    public static Params fromStandardParameters(
+        ValueProvider<ResourceId> baseFilename,
+        @Nullable String shardTemplate,
+        @Nullable String filenameSuffix,
+        boolean windowedWrites) {
+      // Pick the appropriate default policy based on whether windowed writes are being performed.
+      String defaultTemplate =
+          windowedWrites ? DEFAULT_WINDOWED_SHARD_TEMPLATE : DEFAULT_UNWINDOWED_SHARD_TEMPLATE;
+      return new Params(baseFilename,
+          firstNonNull(shardTemplate, defaultTemplate),
+          firstNonNull(filenameSuffix, ""));
+    }
   }
 
   /**
-   * A helper function to construct a {@link DefaultFilenamePolicy} using the standard filename
-   * parameters, namely a provided {@link ResourceId} for the output prefix, and possibly-null
-   * shard name template and suffix.
-   *
-   * <p>Any filename component of the provided resource will be used as the filename prefix.
-   *
-   * <p>If provided, the shard name template will be used; otherwise
-   * {@link #DEFAULT_UNWINDOWED_SHARD_TEMPLATE} will be used for non-windowed file names and
-   * {@link #DEFAULT_WINDOWED_SHARD_TEMPLATE} will be used for windowed file names.
+   */
```

Comment:
@DefaultCoder is a bit awkward to use, since it's designed right now for parameterized coders (e.g. proto and avro). IMO we should make DefaultCoder less awkward for cases like this.